service/api.go

```diff
@@ -159,30 +159,41 @@ func (as *APIServer) requireProject(next http.HandlerFunc) http.HandlerFunc {
 
 func (as *APIServer) requireHost(next http.HandlerFunc) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
-		h, code, err := model.ValidateHost(gimlet.GetVars(r)["hostId"], r)
+		h, _, err := model.ValidateHost(gimlet.GetVars(r)["hostId"], r)
+		UpdateHost(h)
 		if err != nil {
-			as.LoggedError(w, r, code, errors.Wrap(err, "host not assigned to run task"))
-			return
-		}
-		// update host access time
-		if err := h.UpdateLastCommunicated(); err != nil {
-			grip.Warningf("Could not update host last communication time for %s: %+v", h.Id, err)
-		}
-		// Since the host has contacted the app server, we should prevent the
-		// app server from attempting to deploy agents or agent monitors.
-		// Deciding whether or not we should redeploy agents or agent monitors
-		// is handled within the REST route handler.
-		if h.NeedsNewAgent {
-			grip.Warning(message.WrapError(h.SetNeedsNewAgent(false), "problem clearing host needs new agent"))
-		}
-		if h.NeedsNewAgentMonitor {
-			grip.Warning(message.WrapError(h.SetNeedsNewAgentMonitor(false), "problem clearing host needs new agent monitor"))
+			// validate the host again after update in case the agent and app server were out of sync,
+			// and the update fixed it
```

Comment:
I don't understand: how does `UpdateHost` have a bearing on the task the host is assigned?

Reply:
I don't know that it would. I was going off of the comment that `This may mean that "an agent has a different belief about what task it should be running than the app server does."` but it may not actually help at all.

Reply:
the 51 is indeed interesting. I also see 41 though which makes it even more confusing.

Reply:
Thanks to Kim's digging, we have an hypothesis for why it's 51. 
[the agent's client tries 10 times per ](https://github.com/evergreen-ci/evergreen/blob/cda1c49a668203306b3d09a27b1ffc15c62857ce/agent/internal/client/base_client.go#L55-L59)AttachFiles[ call](https://github.com/evergreen-ci/evergreen/blob/cda1c49a668203306b3d09a27b1ffc15c62857ce/agent/internal/client/base_client.go#L55-L59), and maxS3OpAttempts[ can call ](https://github.com/evergreen-ci/evergreen/blob/cda1c49a668203306b3d09a27b1ffc15c62857ce/agent/command/s3_util.go#L16)AttachFiles[ 5 times](https://github.com/evergreen-ci/evergreen/blob/cda1c49a668203306b3d09a27b1ffc15c62857ce/agent/command/s3_util.go#L16) in the outer retry loop for the s3 commands, so 5 * 10 = 50. 
The one extra one is a mystery.

Reply:
nice 🎉